kube-client/src/api/entry.rs

```diff
@@ -0,0 +1,161 @@
+use serde::{de::DeserializeOwned, Serialize};
+
+use crate::{Api, Result};
+
+impl<K: Resource + Clone + DeserializeOwned + Debug> Api<K> {
+    pub async fn entry<'a>(&'a self, name: &'a str) -> Result<Entry<'a, K>> {
+        Ok(match self.try_get(name).await? {
+            Some(object) => Entry::Occupied(OccupiedEntry {
+                api: self,
+                object,
+                dirtiness: Dirtiness::Clean,
+            }),
+            None => Entry::Vacant(VacantEntry { api: self, name }),
+        })
+    }
+}
+
+#[derive(Debug)]
+pub enum Entry<'a, K> {
+    Occupied(OccupiedEntry<'a, K>),
+    Vacant(VacantEntry<'a, K>),
+}
+
+impl<'a, K> Entry<'a, K> {
+    pub fn get(&self) -> Option<&K> {
+        match self {
+            Entry::Occupied(entry) => Some(entry.get()),
+            Entry::Vacant(_) => None,
+        }
+    }
+
+    pub fn get_mut(&mut self) -> Option<&mut K> {
+        match self {
+            Entry::Occupied(entry) => Some(entry.get_mut()),
+            Entry::Vacant(_) => None,
+        }
+    }
+
+    pub fn and_modify(self, f: impl FnOnce(&mut K)) -> Self {
+        match self {
+            Entry::Occupied(entry) => Entry::Occupied(entry.and_modify(f)),
+            entry @ Entry::Vacant(_) => entry,
+        }
+    }
+
+    pub fn or_insert(self, default: impl FnOnce() -> K) -> OccupiedEntry<'a, K>
+    where
+        K: Resource,
+    {
+        match self {
+            Entry::Occupied(entry) => entry,
+            Entry::Vacant(entry) => entry.insert(default()),
+        }
+    }
+}
+
+pub struct OccupiedEntry<'a, K> {
+    api: &'a Api<K>,
+    dirtiness: Dirtiness,
+    object: K,
+}
+
+#[derive(Debug)]
+enum Dirtiness {
+    Clean,
+    Dirty,
+    New,
+}
+
+impl<'a, K> OccupiedEntry<'a, K> {
+    pub fn get(&self) -> &K {
+        &self.object
+    }
+
+    pub fn get_mut(&mut self) -> &mut K {
+        self.dirtiness = match self.dirtiness {
+            Dirtiness::Clean => Dirtiness::Dirty,
+            Dirtiness::Dirty => Dirtiness::Dirty,
+            Dirtiness::New => Dirtiness::New,
+        };
+        &mut self.object
+    }
+
+    pub fn and_modify(mut self, f: impl FnOnce(&mut K)) -> Self {
+        f(self.get_mut());
+        self
+    }
+
+    pub fn into_object(self) -> K {
+        self.object
+    }
+
+    pub async fn sync(&mut self) -> Result<()>
+    where
+        K: Resource + DeserializeOwned + Serialize + Clone + Debug,
+    {
+        self.object = match self.dirtiness {
+            Dirtiness::New => self.api.create(&PostParams::default(), &self.object).await?,
+            Dirtiness::Dirty => {
+                self.api
+                    .replace(
+                        self.object.meta().name.as_deref().unwrap(),
+                        &PostParams::default(),
+                        &self.object,
+                    )
+                    .await?
+            }
+            Dirtiness::Clean => self.api.get(self.object.meta().name.as_deref().unwrap()).await?,
```

Comment:
Not sure about whether we actually want to do this, or just do a NOOP here.

It feels a bit footgunny that we "give up" the optimistic lock based on a runtime condition...

Reply:
On the flip side, if we do implement `get' for `Clean` then we might also want to do the same for `VacantEntry`?

Reply:
To me, this is the big remaining issue before I'd be happy to merge this.

Reply:
The more I think about this, the more I lean towards NOOP. That way, we have a clean flow of "`sync` fails if there are changes, and anyone else has modified the object since the call to `Api::entry`"

Reply:
yeah, I agree. Sync on Clean state should probably be a noop. If people do conditional things to the entry, and then always sync, then we know whether or not it needs to be a noop. I would probably add a one-liner about it for the `sync` fn though.

Reply:
Done. Do we want to rename this to `save`?

Reply:
either is fine with me. slightly leaning towards `commit` because that at least gives the correct impression that we are not fetching upstream changes :thinking:

Reply:
Renamed to `commit`.